File: common/src/main/java/io/xol/chunkstories/content/translator/AbstractContentTranslator.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.content.translator;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.xol.chunkstories.api.content.Content;
import io.xol.chunkstories.api.content.OnlineContentTranslator;
import io.xol.chunkstories.api.content.mods.Mod;
import io.xol.chunkstories.api.entity.Entity;
import io.xol.chunkstories.api.entity.EntityDefinition;
import io.xol.chunkstories.api.item.Item;
import io.xol.chunkstories.api.item.ItemDefinition;
import io.xol.chunkstories.api.net.PacketDefinition;
import io.xol.chunkstories.api.voxel.Voxel;
import io.xol.chunkstories.net.PacketDefinitionImplementation;

/** Assigns IDs for everything that needs one */
public abstract class AbstractContentTranslator implements OnlineContentTranslator {

	final static Logger logger = LoggerFactory.getLogger("content.translator");

	final Content content;
	Set<String> requiredMods;

	Map<Voxel, Integer> voxelMappings;
	private Voxel[] voxelsArray;

	Map<EntityDefinition, Integer> entityMappings;
	private EntityDefinition[] entitiesArray;

	Map<ItemDefinition, Integer> itemMappings;
	private ItemDefinition[] itemsArray;

	Map<PacketDefinition, Integer> packetMappings;
	private PacketDefinition[] packetsArray;

	/** Create an initial content translator */
	public AbstractContentTranslator(Content content) {
		this.content = content;
	}

	public void assignVoxelIds(boolean overwrite) {
		if (overwrite)
			voxelMappings = new HashMap<>();

		content.voxels().all().forEachRemaining(voxel -> {
			if (overwrite || voxelMappings.get(voxel) == null) {
				if (voxel.getName().equals("air"))
					voxelMappings.put(voxel, 0); // Air gets ID 0, always.
				else
					voxelMappings.put(voxel, findNextFreeId(1, voxelMappings.values()));
			}
		});
	}

	public void assignEntityIds(boolean overwrite) {
		if (overwrite)
			entityMappings = new HashMap<>();

		content.entities().all().forEachRemaining(entity -> {
			if (overwrite || entityMappings.get(entity) == null)
				entityMappings.put(entity, findNextFreeId(1, entityMappings.values()));
		});
	}

	public void assignItemIds(boolean overwrite) {
		if (overwrite)
			itemMappings = new HashMap<>();

		content.items().all().forEachRemaining(item -> {
			if (overwrite || itemMappings.get(item) == null)
				itemMappings.put(item, findNextFreeId(1, itemMappings.values()));
		});
	}

	public void assignPacketIds(boolean overwrite) {
		if (overwrite)
			packetMappings = new HashMap<>();

		content.packets().all().forEachRemaining(def -> {
			PacketDefinitionImplementation definition = (PacketDefinitionImplementation) def;

			if (overwrite || packetMappings.get(definition) == null) {

				int packetId = definition.getFixedId();
				if (packetId != -1) {
					packetMappings.put(definition, packetId);
					logger.debug("Using pre-Assignated id " + packetId + " to " + definition.getName());
				}
			}
		});

		content.packets().all().forEachRemaining(def -> {
			PacketDefinitionImplementation definition = (PacketDefinitionImplementation) def;

			if (overwrite || packetMappings.get(definition) == null) {
				int packetId = definition.getFixedId();
				if (packetId == -1) {
					packetId = findNextFreeId(0, packetMappings.values());
					packetMappings.put(definition, packetId);
					logger.debug("Assignated id " + packetId + " to " + definition.getName());
				}
			}
		});
	}

	private int findNextFreeId(int baseId, Collection<Integer> values) {
		int id = baseId;
		while (values.contains(id)) {
			id++;
		}
		return id;
	}

	public void buildArrays() {
		// Create indexable arrays
		voxelsArray = new Voxel[Collections.max(voxelMappings.values()) + 1];
		voxelMappings.forEach((voxel, id) -> voxelsArray[id] = voxel);

		entitiesArray = new EntityDefinition[Collections.max(entityMappings.values()) + 1];
		entityMappings.forEach((entity, id) -> entitiesArray[id] = entity);

		itemsArray = new ItemDefinition[Collections.max(itemMappings.values()) + 1];
		itemMappings.forEach((item, id) -> itemsArray[id] = item);

		if (packetMappings != null && packetMappings.size() > 0) {
			packetsArray = new PacketDefinition[Collections.max(packetMappings.values()) + 1];
			packetMappings.forEach((packet, id) -> packetsArray[id] = packet);
		} else
			packetsArray = new PacketDefinition[0];
	}

	/**
	 * Derives a modified ContentTranslator that takes into account the new content
	 */
	public AbstractContentTranslator loadWith(Content content) {
		return null;
	}

	/**
	 * Internal method to check if a content has the right mods loaded. Returns the
	 * missing mods.
	 */
	Set<String> hasRequiredMods(Content content) {
		Collection<Mod> loadedMods = content.modsManager().getCurrentlyLoadedMods();
		Set<String> loadedModsAsString = new HashSet<>();
		loadedMods.forEach(mod -> loadedModsAsString.add(mod.getModInfo().getInternalName()));

		Set<String> missing = new HashSet<>();

		for (String internalName : requiredMods) {
			if (!loadedModsAsString.contains(internalName))
				missing.add(internalName);
		}

		/*
		 * if(strict) { for(String internalName : loadedModsAsString) {
		 * if(!requiredMods.contains(internalName)) return false; } }
		 */

		return missing;
	}

	/**
	 * Can we load an existing world with the current configuration without issues ?
	 */
	public boolean compatibleWith(Content content) {
		// Check every needed mod is present
		if (hasRequiredMods(content).size() > 0)
			return false;

		// Check every translatable definition has a match
		for (Voxel voxel : voxelMappings.keySet())
			if (content.voxels().getVoxel(voxel.getName()) == null)
				return false;

		for (EntityDefinition entity : entityMappings.keySet())
			if (content.entities().getEntityDefinition(entity.getName()) == null)
				return false;

		for (ItemDefinition item : itemMappings.keySet())
			if (content.items().getItemDefinition(item.getName()) == null)
				return false;

		for (PacketDefinition packet : packetMappings.keySet())
			if (content.packets().getPacketByName(packet.getName()) == null)
				return false;

		return true;
	}

	@Override
	public Collection<String> getRequiredMods() {
		return requiredMods;
	}

	@Override
	public Content getContent() {
		return content;
	}

	@Override
	public int getIdForVoxel(Voxel voxel) {
		return voxelMappings.getOrDefault(voxel, -1);
	}

	@Override
	public Voxel getVoxelForId(int id) {
		if (id < 0 || id >= voxelsArray.length)
			return content.voxels().air();
		return voxelsArray[id];
	}

	@Override
	public int getIdForItem(ItemDefinition definition) {
		return itemMappings.getOrDefault(definition, -1);
	}

	@Override
	public int getIdForItem(Item item) {
		return getIdForItem(item.getDefinition());
	}

	@Override
	public ItemDefinition getItemForId(int id) {
		if (id < 0 || id >= itemsArray.length)
			return null;
		return itemsArray[id];
	}

	@Override
	public int getIdForEntity(EntityDefinition definition) {
		return entityMappings.getOrDefault(definition, -1);
	}

	@Override
	public int getIdForEntity(Entity entity) {
		return getIdForEntity(entity.getDefinition());
	}

	@Override
	public EntityDefinition getEntityForId(int id) {
		if (id < 0 || id >= entitiesArray.length)
			return null;
		return entitiesArray[id];
	}

	@Override
	public int getIdForPacket(PacketDefinition definition) {
		int id = packetMappings.getOrDefault(definition, -1);
		if (id == -1) {
			logger.debug("d:" + definition);
			packetMappings.forEach((d, i) -> logger.debug(d.getName() + ":" + d + ":" + i));
		}
		return id;
	}

	@Override
	public PacketDefinition getPacketForId(int id) {
		if (id < 0 || id >= packetsArray.length)
			return null;
		return packetsArray[id];
	}

	public void write(BufferedWriter writer, boolean writeOnlineStuff) throws IOException {
		for (String internalName : requiredMods) {
			writer.write("requiredMod " + internalName + "\n");
		}

		for (Entry<Voxel, Integer> e : voxelMappings.entrySet()) {
			writer.write("voxel " + e.getValue() + " " + e.getKey().getName() + "\n");
		}

		for (Entry<ItemDefinition, Integer> e : itemMappings.entrySet()) {
			writer.write("item " + e.getValue() + " " + e.getKey().getName() + "\n");
		}

		for (Entry<EntityDefinition, Integer> e : entityMappings.entrySet()) {
			writer.write("entity " + e.getValue() + " " + e.getKey().getName() + "\n");
		}

		if (writeOnlineStuff) {
			for (Entry<PacketDefinition, Integer> e : packetMappings.entrySet()) {
				writer.write("packet " + e.getValue() + " " + e.getKey().getName() + "\n");
			}
		}
	}

	public String toString() {
		return toString(true);
	}

	public String toString(boolean writeOnlineStuff) {
		StringBuilder sb = new StringBuilder();

		for (String internalName : requiredMods) {
			sb.append("requiredMod " + internalName + "\n");
		}

		for (Entry<Voxel, Integer> e : voxelMappings.entrySet()) {
			sb.append("voxel " + e.getValue() + " " + e.getKey().getName() + "\n");
		}

		for (Entry<ItemDefinition, Integer> e : itemMappings.entrySet()) {
			sb.append("item " + e.getValue() + " " + e.getKey().getName() + "\n");
		}

		for (Entry<EntityDefinition, Integer> e : entityMappings.entrySet()) {
			sb.append("entity " + e.getValue() + " " + e.getKey().getName() + "\n");
		}

		if (writeOnlineStuff) {
			for (Entry<PacketDefinition, Integer> e : packetMappings.entrySet()) {
				sb.append("packet " + e.getValue() + " " + e.getKey().getName() + "\n");
			}
		}

		return sb.toString();
	}

	public void test() {
		try {
			BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
			write(writer, true);
			writer.flush();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	public void save(File file) {
		try {
			file.getParentFile().mkdirs();
			BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(file)));
			write(writer, false);
			writer.flush();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

}

File: common/src/main/java/xyz/chunkstories/content/translator/AbstractContentTranslator.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//
package xyz.chunkstories.content.translator

import org.slf4j.LoggerFactory
import xyz.chunkstories.api.block.BlockType
import xyz.chunkstories.api.content.Content
import xyz.chunkstories.api.content.ContentTranslator
import xyz.chunkstories.api.content.mods.Mod
import xyz.chunkstories.api.entity.Entity
import xyz.chunkstories.api.entity.EntityDefinition
import xyz.chunkstories.api.item.Item
import xyz.chunkstories.api.item.ItemDefinition
import xyz.chunkstories.api.net.PacketId
import xyz.chunkstories.content.GameContentStore
import xyz.chunkstories.net.PacketDefinition
import java.io.*
import java.lang.StringBuilder
import java.util.*
import kotlin.collections.HashSet

/** Assigns IDs for everything that needs one  */
abstract class AbstractContentTranslator(override val content: GameContentStore) : ContentTranslator {
    override var requiredMods: MutableSet<String> = mutableSetOf()
        protected set

    var voxelMappings: MutableMap<BlockType, Int> = mutableMapOf()
    private lateinit var voxelsArray: Array<BlockType?>

    var entityMappings: MutableMap<EntityDefinition, Int> = mutableMapOf()
    private lateinit var entitiesArray: Array<EntityDefinition?>

    var itemMappings: MutableMap<ItemDefinition, Int> = mutableMapOf()
    private lateinit var itemsArray: Array<ItemDefinition?>

    var packetMappings: MutableMap<PacketDefinition, Int> = mutableMapOf()
    private lateinit var packetsArray: Array<PacketDefinition?>

    fun assignVoxelIds(overwrite: Boolean) {
        if (overwrite)
            voxelMappings = mutableMapOf()
        content.blockTypes.all.forEach { blockType ->
            if (voxelMappings[blockType] == null) {
                if (blockType.name == "air")
                    voxelMappings[blockType] = 0 // Air gets ID 0, always.
                else
                    voxelMappings[blockType] = findNextFreeId(1, voxelMappings.values)
            }
        }
    }

    fun assignEntityIds(overwrite: Boolean) {
        if (overwrite)
            entityMappings = mutableMapOf()
        content.entities.all.forEach {
            entity -> if (entityMappings[entity] == null)
                entityMappings[entity] = findNextFreeId(1, entityMappings!!.values)
        }
    }

    fun assignItemIds(overwrite: Boolean) {
        if (overwrite) itemMappings = mutableMapOf()
        content.items.all.forEach { item ->
            if (overwrite || itemMappings[item] == null)
                itemMappings[item] = findNextFreeId(1, itemMappings.values)
        }
    }

    fun assignPacketIds(overwrite: Boolean) {
        if (overwrite) packetMappings = mutableMapOf()
        content.packets.all.forEach { def ->
            if (overwrite || packetMappings[def] == null) {
                val packetId: Int = def.fixedId
                if (packetId != -1) {
                    packetMappings[def] = packetId
                    logger.debug("Using known assigned id " + packetId + " for " + def.name)
                }
            }
        }
        content.packets.all.forEach{ def ->
            if (overwrite || packetMappings[def] == null) {
                var packetId: Int = def.fixedId
                if (packetId == -1) {
                    packetId = findNextFreeId(0, packetMappings!!.values)
                    packetMappings[def] = packetId
                    logger.debug("Assigned id " + packetId + " to " + def.name)
                }
            }
        }
    }

    private fun findNextFreeId(baseId: Int, values: Collection<Int>): Int {
        var id = baseId
        while (values.contains(id)) {
            id++
        }
        return id
    }

    fun buildArrays() {
        // Create indexable arrays
        voxelsArray = arrayOfNulls(Collections.max(voxelMappings.values) + 1)
        voxelMappings.forEach { (voxel: BlockType, id: Int) -> voxelsArray[id] = voxel }
        entitiesArray = arrayOfNulls(Collections.max(entityMappings.values) + 1)
        entityMappings.forEach { (entity: EntityDefinition?, id: Int?) -> entitiesArray[id] = entity }
        itemsArray = arrayOfNulls(Collections.max(itemMappings.values) + 1)
        itemMappings.forEach { (item, id) -> itemsArray[id] = item }
        if (packetMappings.size > 0) {
            packetsArray = arrayOfNulls(Collections.max(packetMappings.values) + 1)
            packetMappings.forEach { (packet: PacketDefinition?, id: Int?) -> packetsArray[id] = packet }
        } else packetsArray = arrayOfNulls(0)
    }

    /**
     * Derives a modified ContentTranslator that takes into account the new content
     */
    fun loadWith(content: Content?): AbstractContentTranslator? {
        TODO("was this ever used")
    }

    /**
     * Internal method to check if a content has the right mods loaded. Returns the
     * missing mods.
     */
    fun hasRequiredMods(content: Content): Set<String> {
        val loadedMods = content.modsManager.currentlyLoadedMods
        val loadedModsAsString: MutableSet<String> = HashSet()
        loadedMods.forEach {
            mod: Mod -> loadedModsAsString.add(mod.modInfo.internalName)
        }
        val missing: MutableSet<String> = HashSet()
        for (internalName in requiredMods) {
            if (!loadedModsAsString.contains(internalName)) missing.add(internalName)
        }

        return missing
    }

    /**
     * Can we load an existing world with the current configuration without issues ?
     */
    fun compatibleWith(content: GameContentStore): Boolean {
        // Check every needed mod is present
        if (hasRequiredMods(content).isNotEmpty()) return false

        // Check every translatable definition has a match
        for (voxel in voxelMappings.keys) if (content.blockTypes[voxel.name] == null) return false
        for (entity in entityMappings.keys) if (content.entities.getEntityDefinition(entity.name) == null) return false
        for (item in itemMappings.keys) if (content.items.getItemDefinition(item.name) == null) return false
        for (packet in packetMappings.keys) if (content.packets.getPacketByName(packet.name) == null) return false
        return true
    }

    override fun getIdForVoxel(voxel: BlockType): Int {
        return voxelMappings.getOrDefault(voxel, -1)
    }

    override fun getVoxelForId(id: Int): BlockType? {
        return if (id >= 0 && id < voxelsArray.size) voxelsArray[id] else null
    }

    override fun getIdForItem(definition: ItemDefinition): Int {
        return itemMappings.getOrDefault(definition, -1)
    }

    override fun getIdForItem(item: Item?): Int {
        return getIdForItem(item!!.definition)
    }

    override fun getItemForId(id: Int): ItemDefinition? {
        return if (id < 0 || id >= itemsArray.size) null else itemsArray[id]
    }

    override fun getIdForEntity(definition: EntityDefinition): Int {
        return entityMappings.getOrDefault(definition, -1)
    }

    override fun getIdForEntity(entity: Entity): Int {
        return getIdForEntity(entity.definition)
    }

    override fun getEntityForId(id: Int): EntityDefinition? {
        return if (id < 0 || id >= entitiesArray.size) null else entitiesArray[id]
    }

    fun getIdForPacket(definition: PacketDefinition): PacketId {
        val id = packetMappings.getOrDefault(definition, -1)
        if (id == -1) {
            logger.debug("d:$definition")
            packetMappings.forEach { (def, i) ->
                logger.debug(def.name + ":" + def + ":" + i)
            }
        }
        return id
    }

    fun getPacketForId(id: PacketId): PacketDefinition? {
        return if (id < 0 || id >= packetsArray.size) null else packetsArray[id]
    }

    fun write(writer: BufferedWriter, writeOnlineStuff: Boolean) {
        writer.write(toString(writeOnlineStuff))
    }

    override fun toString(): String {
        return toString(true)
    }

    fun toString(writeOnlineStuff: Boolean): String {
        val sb = StringBuilder()
        for (internalName in requiredMods) {
            sb.append("requiredMod $internalName\n")
        }
        for ((key, value) in voxelMappings) {
            sb.append("voxel $value ${key.name}\n")
        }
        for ((key, value) in itemMappings) {
            sb.append("item $value ${key.name}\n")
        }
        for ((key, value) in entityMappings) {
            sb.append("entity $value ${key.name}\n")
        }
        if (writeOnlineStuff) {
            for ((key, value) in packetMappings) {
                sb.append("packet $value ${key.name}\n")
            }
        }
        return sb.toString()
    }

    fun test() {
        try {
            val writer = BufferedWriter(OutputStreamWriter(System.out))
            write(writer, true)
            writer.flush()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun save(file: File) {
        try {
            file.parentFile.mkdirs()
            val writer = BufferedWriter(OutputStreamWriter(FileOutputStream(file)))
            write(writer, false)
            writer.flush()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    companion object {
        val logger = LoggerFactory.getLogger("content.translator")
    }

}